### ISSUE TITLE
"NonExists" could not be found in the located source  

### ISSUE THREAD

**Issue by POPSuL** [ApiGen/ApiGen#929]
Code:
```
class B extends NonExists
{
}
```

Actual result:
```
% php ./bin/apigen generate -vvv --destination=/var/www/api ./test                                                                 

                                                                                                          
  [Roave\BetterReflection\Reflector\Exception\IdentifierNotFound]                                         
  Roave\BetterReflection\Reflection\ReflectionClass "NonExists" could not be found in the located source  
                                                                                                          

Exception trace:
 () at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflector/Exception/IdentifierNotFound.php:31
 Roave\BetterReflection\Reflector\Exception\IdentifierNotFound::fromIdentifier() at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflector/ClassReflector.php:55
 Roave\BetterReflection\Reflector\ClassReflector->reflect() at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflection/ReflectionClass.php:732
 Roave\BetterReflection\Reflection\ReflectionClass->getParentClass() at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflection/ReflectionClass.php:1141
 Roave\BetterReflection\Reflection\ReflectionClass->getInheritanceClassHierarchy() at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflection/ReflectionClass.php:971
 Roave\BetterReflection\Reflection\ReflectionClass->getInterfaces() at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflection/ReflectionClass.php:999
 Roave\BetterReflection\Reflection\ReflectionClass->getInterfaceNames() at /home/popsul/projects/ApiGen/vendor/roave/better-reflection/src/Reflection/ReflectionClass.php:1061
 Roave\BetterReflection\Reflection\ReflectionClass->implementsInterface() at /home/popsul/projects/ApiGen/packages/Reflection/src/Reflection/Class_/ClassReflection.php:177
 ApiGen\Reflection\Reflection\Class_\ClassReflection->implementsInterface() at /home/popsul/projects/ApiGen/packages/Element/src/ReflectionCollector/NamespaceReflectionCollector.php:45
 ApiGen\Element\ReflectionCollector\NamespaceReflectionCollector->processReflection() at /home/popsul/projects/ApiGen/packages/Element/src/ReflectionCollectorCollector.php:24
 ApiGen\Element\ReflectionCollectorCollector->processReflection() at /home/popsul/projects/ApiGen/packages/Reflection/src/TransformerCollector.php:76
 ApiGen\Reflection\TransformerCollector->transformSingle() at /home/popsul/projects/ApiGen/packages/Reflection/src/TransformerCollector.php:42
 ApiGen\Reflection\TransformerCollector->transformGroup() at /home/popsul/projects/ApiGen/packages/Reflection/src/Parser/Parser.php:98
 ApiGen\Reflection\Parser\Parser->transformBetterClassInterfaceAndTraitReflections() at /home/popsul/projects/ApiGen/packages/Reflection/src/Parser/Parser.php:104
 ApiGen\Reflection\Parser\Parser->parseClassElements() at /home/popsul/projects/ApiGen/packages/Reflection/src/Parser/Parser.php:54
 ApiGen\Reflection\Parser\Parser->parseDirectories() at /home/popsul/projects/ApiGen/src/Application/ApiGenApplication.php:55
 ApiGen\Application\ApiGenApplication->runCommand() at /home/popsul/projects/ApiGen/src/Console/Command/GenerateCommand.php:61
 ApiGen\Console\Command\GenerateCommand->execute() at /home/popsul/projects/ApiGen/vendor/symfony/console/Command/Command.php:264
 Symfony\Component\Console\Command\Command->run() at /home/popsul/projects/ApiGen/vendor/symfony/console/Application.php:869
 Symfony\Component\Console\Application->doRunCommand() at /home/popsul/projects/ApiGen/vendor/symfony/console/Application.php:223
 Symfony\Component\Console\Application->doRun() at /home/popsul/projects/ApiGen/vendor/symfony/console/Application.php:130
 Symfony\Component\Console\Application->run() at /home/popsul/projects/ApiGen/bin/apigen:16

generate [--destination DESTINATION] [--] <source> (<source>)...
```

Expected behavior:
Documentation should be generated correctly, but without anchor to `NonExists` class.

**Comment by nelson6e65** [comment]
I think you are kind of right about throwing the exception, but... You can't generate an incomplete API of an incomplete class. You need to know what B overrides from base class.

**Comment by POPSuL** [comment]
@nelson6e65 yes, we need information about inherited methods. But what if I have `src` with 2 files which extends/implements some class/interface from `vendor`? I need launch apigen with path to `src/..` to generate complete documentation, but it will contain documentation for 2 my classes + documentation for 5-10-100 vendor libraries with own tests. This is wrong behavior.

**Comment by POPSuL** [comment]
Now I launched `php ./bin/apigen generate -vvv --destination=/var/www/api /var/www/core` to generate complete documentation (src, tests, vendor + some bootstrap files), and... 
15 minutes have passed. php use 100% CPU and 2GB of RAM. Nothing is output to the console...

![image](https://user-images.githubusercontent.com/683358/27850500-dfb58842-61a0-11e7-8502-d165ba931a2e.png)

**Comment by POPSuL** [comment]
And after ~25 min. I get Exception for vendor code (explained here https://github.com/ApiGen/ApiGen/issues/864).

**Comment by vlastavesely** [comment]
I think that it *should* fail for really missing classes. But maybe with some prettier error message...

On support to parse `vendor` classes we already work in #916 

If you generate API from `/path/to/project/src/`, ApiGen is going to check whether directory `/path/to/project/vendor` exists and if so, composer loader is used to load vendor classes.

**Comment by TomasVotruba** [comment]
ApiGen should fail on missing classes, as works only with code itself working. Making it missing-code-proof would require lot of work which massively increase complexity with not much added value. Use case is working code, this is not static analysis.

> On support to parse vendor classes we already work in #916

That was merged to master. @POPSuL Could you try with that for /vendor dependencies?

**Comment by TomasVotruba** [comment]
Any updates?

**Comment by POPSuL** [comment]
For my case: #930 didnt solved this problem.
I have 
```
 % php ./bin/apigen generate -vvv --destination=/var/www/api /var/www/core/src /var/www/core/system.php /var/www/core/exceptions.php /var/www/core/helpers.php


  [Roave\BetterReflection\Reflector\Exception\IdentifierNotFound]
  Roave\BetterReflection\Reflection\ReflectionClass "ApplicationException" could not be found in the located source
```

class ApplicationException declared to root ns in `exceptions.php`, but i got exception :(
All mentionings of ApplicationException:

```
[popsul] ~/projects/ApiGen [test-branch] ?? % grep -HR 'ApplicationException' /var/www/core
/var/www/core/exceptions.php:class ApplicationException extends \BaseException
/var/www/core/clover.xml:      <class name="ApplicationException" namespace="global">
/var/www/core/src/Api/Messaging/Mail/MailView.php:     * @throws \ApplicationException
/var/www/core/src/Api/Messaging/Mail/MailView.php:                throw new \ApplicationException(sprintf('mailview %s not found', $file));
/var/www/core/src/Api/Messaging/Telegram/Exception/ValidationException.php:class ValidationException extends \ApplicationException
/var/www/core/src/Application/Exception/HttpException.php:class HttpException extends \ApplicationException
/var/www/core/src/Core/Db/Orm/AbstractDbRepository.php:use ApplicationException;
/var/www/core/src/Core/Db/Orm/AbstractDbRepository.php:     * @throws ApplicationException если репозиторий хранит в себе модели с составным ключом
/var/www/core/src/Core/Db/Orm/AbstractDbRepository.php:            throw new ApplicationException('Метод findByPks не поддерживается для моделей с составным ключом');
/var/www/core/tests/Core/Db/Orm/AbstractDbRepositoryTest.php:            $this->fail('Исключение ApplicationException о недоступности метода не выброшено');
/var/www/core/tests/Core/Db/Orm/AbstractDbRepositoryTest.php:        } catch (\ApplicationException $e) {
```

**Comment by TomasVotruba** [comment]
Could you send a failing test then, please

**Comment by TomasVotruba** [comment]
Wait, is that non existing class or non-autoloaded class?

Non existing classes are not supported, as ApiGen works only with working code.

If that is non-autoloaded class, do you load it in `composer.json`?
ApiGen doesn't aim to replace composer.

**Comment by POPSuL** [comment]
@TomasVotruba it is non-autoloaded class. ApplicationException defined in `exceptions.php`. Invoke command in previous comment.

**Comment by POPSuL** [comment]
If I remove all sources except /var/www/core/src - I get same exception, but in that case class really been not defined (because apigen nothing known about `exceptions.php` with `ApplicationException`)

**Comment by TomasVotruba** [comment]
@POPSuL Just add it to `composer.json` then:

```json
{
    "autoload": {
        "files": ["src/exceptions.php"]
    }
}
```

and dump autoload:

```bash
composer dump
```

**Comment by vlastavesely** [comment]
It waits on an external PR? On which?

After #957 it could be working if there is not another problem.

@POPSuL could you try it?

**Comment by POPSuL** [comment]
@vlastavesely I'll try tomorrow

**Comment by pv4** [comment]
> Non-existing classes are not supported

@TomasVotruba, could you please reconsider this decision?

I can only speak for the company I work at, but there are many heavyweight (tens or sometimes hundreds megabytes of ugly php code) cms/frameworks and they are often not documented at all. When we develop modules for such systems we'd like to document our code using phpDoc.

Trying to run ApiGen for the whole system is pointless. That would only spend cpu time and eat ram with no actual gain.

Aside form technical reasons, from business point of view we don't have to provide documentation for the framework we use. The framework whether is documented or is not, while we are responsive for our own code only.

I understand that ideally ApiGen has to know about parent classes to generate proper documentation. But I think it would be great to have a workaround option like --ignore-unknown-classes which assumes every non-existing class to be an empty extends-less, implements-less class like
```
class NonFoundClass {
}
```
Is it possible to implement this?

Also AFAIU ApiGen 4.1.2 handles unknown classes fine so this issue is clearly a regression.

**Comment by TomasVotruba** [comment]
@POPSuL Any updates?

@pv4 I think it was successfully resolved 2-3 comments above. 
Is it still broken with current master?

**Comment by POPSuL** [comment]
If add file to `"autoload"` section in composer.json - all works properly.
But option `--ignore-unknown-classes` should be implemented.
Use cases:
* If we use old library which added via `include` instead of `composer`
* If we use specific native extension (e.g. we have some native extension on production servers, but not in build server)

I agree with @pv4 - this is regression.

**Comment by pv4** [comment]
@TomasVotruba

My case is that some (hundreds of) classes are **non-existing**. The fix above is supposed to fix **non-autoloaded** classes.

For my case the current master outputs
```
  [Roave\BetterReflection\Reflector\Exception\IdentifierNotFound]                                                      
  Roave\BetterReflection\Reflection\ReflectionClass "Non\Existing\Class\Name" could not be found in the located source  
```

**Comment by TomasVotruba** [comment]
@POPSuL Thanks for updates.

All right, looks valid to me.

Do you have idea how to implement this nicely?

**Comment by pv4** [comment]
I haven't looked at the ApiGen source code so I cannot say how this can be better implemented from the developer point of view.

As a user, I have not deep understanding what features ApiGen has and what exactly depends on this issue. I think "ignore" is the best strategy for "unknown" identifiers: namespaces, classes, interfaces, traits, methods, variables, constants... So, if a class extends an unknown class then assume the parent class is empty. If it implements an unknown interface then assume it's an empty interface. If some @see mentions an unknown entity, don't make it a link, leave it a plain text. Etc...

That is enough to close the regression. That is enough for my needs. And I think that is logical enough.

A disablable (from command like) warning would be fine here. The option can be something like --unknown=fail (the current behavior), --unknown=warn (ignore unknown entities but print a warning, so a user can verify he haven't missed anything important), --unknown=ignore (ignore unknown entities without a warning so that other useful messages are not lost among hundreds of warnings).

--

The probable more advanced way is suggested below. But it's just a first thought. It's not required for me currently, it's probably **too hard to implement**, and there will still be people whose needs are not covered by it. It's here just FYI.

The solution is some mapping from entity names to "config" php-files containing the "template" definitions. Imagine we write something like
```
apigen --map-def "\External\DbLibrary\.*" --map-to /tmp/classTemplate.php \
       --map-def ".*\Interfaces\.*" --map-to /tmp/interfaceTemplate.php \
       --map-def "\Library\Helpers\.*" --map-to /tmp/helperTemplate.php
```
and have the files:
```
// classTemplate.php
class * {}
```
```
// interfaceTemplate.php
interface * {}
```
```
// helperTemplate.php
class * extends \Library\Helpers\Base {
public function __construct ($someVar) {}
}
```
Then this could mean:
1. anything that is referred as \External\DbLibrary\<anything> must be treated an empty class,
2. anytiong that is referred as <anything>\Interfaces\<anything> must be treated as an empty interface,
3. anything that is referred as \Library\Helpers\<anything> must be treated as a class that extends \Library\Helpers\Base and has a constructor with a single argument.

**Comment by TomasVotruba** [comment]
Would be able to put that into PR with bit of help?

**Comment by TomasVotruba** [comment]
Also, please create a new issue. There are various things mixed up, so I'm closing it.

**Comment by pv4** [comment]
> That would be needed, if you want to add this feature.

Ok, now I understand you meant it's up to me to implement the feature :)

Unfortunatelly I have no enough time currently and cannot predict when I have the time to implement what I need by myself (and that definitely won't happen in the nearest 2-3 months anyway, so please do NOT delay the 5.0.0 release just because of waiting for me).

> Would you like to do that with my help?

Thanks, I'll ask when I have anything to ask. I think I would understand the architecture (and possible ways to solve the issue) better if I explore the sources by myself so I'll try to go this way.

> Also, please create a new issue.

Create the issue NOW? I was going to create an issue with pull request as soon as it's ready.
